src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
+            private readonly ArrayBuilder<BoundExpression> _sideEffectBuilder;
+            private readonly DagTempAllocator _tempAllocator;
 
-                case BoundKind.ConstantPattern:
-                    {
-                        var constantPattern = (BoundConstantPattern)loweredPattern;
-                        return MakeIsConstantPattern(constantPattern, loweredInput);
-                    }
+            private readonly BoundExpression _loweredInput;
+            private readonly BoundDagTemp _inputTemp;
 
-                default:
-                    throw ExceptionUtilities.UnexpectedValue(loweredPattern.Kind);
+            public IsPatternTranslator(LocalRewriter localRewriter, BoundExpression loweredInput)
+            {
+                this._localRewriter = localRewriter;
+                this._factory = localRewriter._factory;
+                this._tempAllocator = new DagTempAllocator(localRewriter._factory);
+                this._loweredInput = loweredInput;
+                this._inputTemp = new BoundDagTemp(loweredInput.Syntax, loweredInput.Type, null, 0);
+                this._conjunctBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+                this._sideEffectBuilder = ArrayBuilder<BoundExpression>.GetInstance();
             }
-        }
 
-        BoundPattern LowerPattern(BoundPattern pattern)
-        {
-            switch (pattern.Kind)
+            public void Dispose()
             {
-                case BoundKind.DeclarationPattern:
-                    {
-                        var declPattern = (BoundDeclarationPattern)pattern;
-                        return declPattern.Update(declPattern.Variable, VisitExpression(declPattern.VariableAccess), declPattern.DeclaredType, declPattern.IsVar);
-                    }
-                case BoundKind.ConstantPattern:
-                    {
-                        var constantPattern = (BoundConstantPattern)pattern;
-                        return constantPattern.Update(VisitExpression(constantPattern.Value), constantPattern.ConstantValue);
-                    }
-                default:
-                    return pattern;
+                _conjunctBuilder.Free();
+                _sideEffectBuilder.Free();
+                _tempAllocator.Dispose();
             }
-        }
 
-        private BoundExpression MakeIsConstantPattern(BoundConstantPattern loweredPattern, BoundExpression loweredInput)
-        {
-            return CompareWithConstant(loweredInput, loweredPattern.Value);
-        }
+            public class DagTempAllocator : IDisposable
+            {
+                private readonly SyntheticBoundNodeFactory _factory;
+                private readonly PooledDictionary<BoundDagTemp, BoundExpression> _map = PooledDictionary<BoundDagTemp, BoundExpression>.GetInstance();
+                private readonly ArrayBuilder<LocalSymbol> _temps = ArrayBuilder<LocalSymbol>.GetInstance();
```

Comment:
I'm surprised that the temps don't need a lifetime. When do we know that a temp is dead?

Reply:
All of these temps are alive for the duration of the entire pattern-matching machinery. Sharing them is hard (for a switch or match), but a problem to be solved later.

---
In reply to: [155096500](https://github.com/dotnet/roslyn/pull/23209#discussion_r155096500) [](ancestors = 155096500)